### PR TITLE
Added NASM grammar

### PR DESCRIPTION
**Pull request by Caralludo**

### CODE REVIEW

File: asm/nasm/nasm_x86_64_Lexer.g4 [hunk 1]
```diff
@@ -0,0 +1,2702 @@
...
+    | '0d' [0-9][0-9_]*
+    ;
+
+SIGN
+    : PLUS | MINUS
+    ;
+
+OCT_INTEGER
+    : [0-7][0-7_]* ('q' | 'o')
+    | ('0q' | '0o') [0-7][0-7_]*
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Lexer.g4:2644]:
```suggestion
    | '0' [qo][0-7][0-7_]*
```

File: asm/nasm/nasm_x86_64_Lexer.g4 [hunk 2]
```diff
@@ -0,0 +1,2702 @@
...
+    ;
+
+HEX_INTEGER
+    : [0-9A-F][0-9A-F_]+ 'h'
+    | '$' [0-9A-F][0-9A-F_]*
+    | ('0x' | '0h') [0-9A-F][0-9A-F_]*
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Lexer.g4:2650]:
```suggestion
    | '0' [xh][0-9A-F][0-9A-F_]*
```

File: asm/nasm/nasm_x86_64_Lexer.g4 [hunk 3]
```diff
@@ -0,0 +1,2702 @@
...
+    ;
+
+BIN_INTEGER
+    : [01][01_]+ ('b' | 'y')
+    | ('0b' | '0y') [01][01_]*
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Lexer.g4:2655]:
```suggestion
    | '0' [by][01][01_]*
```

File: asm/nasm/nasm_x86_64_Lexer.g4 [hunk 4]
```diff
@@ -0,0 +1,2702 @@
...
+WARNING_NAME
+    : NAME (MINUS NAME)+
+    ;
+
+NAME
+    : [a-z._?][a-z0-9_$#@~.?]*
+    ;
+
+PREPROCESSOR_DIRECTIVES
+    : ('%define' | '%xdefine' | '%ixdefine' | '%undef' | '%assign' | '%iassign' | '%defstr' | '%deftok' | '%defalias' | '%undefalias' | '%strcat' | '%strlen' | '%substr' | '%include' | '%pathsearch' | '%depend' | '%use' | '%line' | '%clear') ~[\r\n]*
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Lexer.g4:2680]:
```suggestion
    : '%' ('define' | 'xdefine' | 'ixdefine' | 'undef' | 'assign' | 'iassign' | 'defstr' | 'deftok' | 'defalias' | 'undefalias' | 'strcat' | 'strlen' | 'substr' | 'include' | 'pathsearch' | 'depend' | 'use' | 'line' | 'clear') ~[\r\n]*
```

File: asm/nasm/nasm_x86_64_Parser.g4 [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
+    caseInsensitive=true;
+    tokenVocab=nasm_x86_64_Lexer;
+}
+
+program
+    : line* EOF
+    ;
+
+line
+    : label? (directive | times_prefix? pseudoinstruction | times_prefix? instruction)? EOL
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Parser.g4:13]:
```suggestion
    : label? (directive | times_prefix? (pseudoinstruction | instruction))? EOL
```

File: asm/nasm/nasm_x86_64_Parser.g4 [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+
+directive
+    : bits decimal_integer
+    | use16
+    | use32
+    | default default_perfix
+    | section section_params
+    | absolute (integer | name)
+    | extern extern_params
+    | required extern_params
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Parser.g4:28]:
```suggestion
    | (extern | required) extern_params
```

File: asm/nasm/nasm_x86_64_Parser.g4 [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+    | global global_params
+    | common common_params
+    | static name
+    | cpu (decimal_integer | name)
+    | float_name float_params
+    | LEFT_BRACKET warning warning_state? warning_class RIGHT_BRACKET
+    | org integer
+    | LEFT_BRACKET map map_type name RIGHT_BRACKET
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Parser.g4:36]:
```suggestion
    | LEFT_BRACKET (warning warning_state? warning_class | map map_type name) RIGHT_BRACKET
    | org integer
```

File: asm/nasm/nasm_x86_64_Parser.g4 [hunk 8]
```diff
@@ -0,0 +1,553 @@
...
+    | group name name name+
+    | uppercase
+    | import_rule name+
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Parser.g4:39]:
```suggestion
    | (group name name | import_rule) name+
    | uppercase
```

File: asm/nasm/nasm_x86_64_Parser.g4 [hunk 9]
```diff
@@ -0,0 +1,553 @@
...
+    | NOWRITE
+    ;
+
+starting_possition
+    : VSTART EQUAL_1 integer
+    ;
+
+follow
+    : FOLLOWS EQUAL_1 name
+    | VFOLLOWS EQUAL_1 name
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Parser.g4:145]:
```suggestion
    : (FOLLOWS | VFOLLOWS) EQUAL_1 name
```

File: asm/nasm/nasm_x86_64_Parser.g4 [hunk 10]
```diff
@@ -0,0 +1,553 @@
...
+    : FLAT
+    ;
+
+absolute_seg
+    : ABSOLUTE EQUAL_1 integer
+    ;
+
+alingment
+    : ALIGN EQUAL_1 integer
+    | START EQUAL_1 integer
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Parser.g4:158]:
```suggestion
    : (ALIGN | START) EQUAL_1 integer
```

File: asm/nasm/nasm_x86_64_Parser.g4 [hunk 11]
```diff
@@ -0,0 +1,553 @@
...
+times
+    : TIMES
+    ;
+
+
+pseudoinstruction
+    : name? dx value (COMMA value)*
+    | name? resx integer
+    | name? incbin atom (COMMA atom)*
+    | name? equ (integer | expression)
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Parser.g4:363]:
```suggestion
    : name? ( dx value (COMMA value)*
            | resx integer
            | incbin atom (COMMA atom)*
            | equ (integer | expression)
            )
```

File: asm/nasm/nasm_x86_64_Parser.g4 [hunk 12]
```diff
@@ -0,0 +1,553 @@
...
+
+strict
+    : STRICT
+    ;
+
+macro_call
+    : name macro_param? (COMMA macro_param)*
+    | name LEFT_PARENTHESIS macro_param? (COMMA macro_param)* RIGHT_PARENTHESIS
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Parser.g4:543]:
```suggestion
    : name ( macro_param? (COMMA macro_param)*
           | LEFT_PARENTHESIS macro_param? (COMMA macro_param)* RIGHT_PARENTHESIS
           )
```

**Review by KvanTTT — CHANGES_REQUESTED** [review]
Thanks for the new grammar but please simplify code in some places.

Feel free to argue why it's not possible (and add an explaining comment).

File: asm/nasm/nasm_x86_64_Parser.g4 [hunk 13]
```diff
@@ -0,0 +1,550 @@
...
+
+operand
+    : (register COLON)? register
+    | (strict? size)? string
+    | (strict? size)? float_number
+    | (strict? size)? integer
+    | (register COLON)? name
+    | expression
+    | (strict? size)? LEFT_BRACKET expression (COMMA expression)* RIGHT_BRACKET
```
Comment (KvanTTT) [inline asm/nasm/nasm_x86_64_Parser.g4:528]:
```suggestion
    : (register COLON)? (register | name)
    | (strict? size)? (string | float_number | integer | LEFT_BRACKET expression (COMMA expression)* RIGHT_BRACKET)
    | expression
```

**Review by KvanTTT — APPROVED** [review]
Thanks a lot for the fixes!

**Comment by Caralludo** [comment]
Sorry for the inconveniences, I made a lot of rooky errors.

**Pull request merged**